message-queue-lib/src/cmr/message_queue/queue/sqs.clj

```diff
@@ -132,18 +132,26 @@
                       ;; Tell SQS how long to wait before returning with no data (long polling).
                       (.setWaitTimeSeconds (Integer. (queue-polling-timeout))))]
     (a/thread
-      (try
-        (u/while-let [rec-result (.receiveMessage sqs-client rec-req)]
-          (when-let [msg (first (.getMessages rec-result))]
-            (let [msg-body (.getBody msg)
-                  msg-content (json/decode msg-body true)]
-              (try
-                (handler msg-content)
-                (.deleteMessage sqs-client queue-url (.getReceiptHandle msg))
-                (catch Throwable e
-                  (error e "Message processing failed for message" (pr-str msg) "on queue" queue-name))))))
-        (catch Throwable e
-          (error  e "Async handler for queue" queue-name "completing."))))))
+      (loop []
+        (try
+          (let [rec-result (.receiveMessage sqs-client rec-req)]
+            (when-let [msg (first (.getMessages rec-result))]
+              (let [msg-body (.getBody msg)
+                    msg-content (json/decode msg-body true)]
+                (try
+                  (handler msg-content)
+                  (.deleteMessage sqs-client queue-url (.getReceiptHandle msg))
+                  (catch Throwable e
+                    (error e "Message processing failed for message" (pr-str msg) "on queue" queue-name))))))
+          (catch Exception e
+            (error  e "Async handler for queue" queue-name "continuing after failed message receive."))
+          ;; Catching this just so we can log the thread exiting so it will show up in splunk.
+          (catch Throwable t
```

Comment:
Agreed -- that's exactly how I've done custom "exceptions" in Clojure in previous projects. I much prefer the ex-info data to a Java stacktrace any day ...